### PR TITLE
Instantiate this when used only in type parameter constraints

### LINKED ISSUE #16930: Func with param of type 'keyof this' doesn't  resolve property of inherited class
**TypeScript Version:**  2.4.0 

**Code**

```ts
class base {
  public baseProp;
    check<TProp extends keyof this>(prop: TProp): boolean {
        return true;
    }
}

class test extends base {
    public inherited: string;

    constructor() {
      super();
        this.check('baseProp'); //ok
        this.check('inherited'); // error 'inherited' is not assignable to parameter of type keyof this
    }    
}

```

**Expected behavior:**
no error

### PR DESCRIPTION
**Pull request by sandersn**
Check function type parameters for this references. Previously, when checking for this references, the compiler did not check type parameters. This caused it to miss some instantiations that were necessary.

Also some cleanup:
1. Rename isIndependent* to is*FreeOfThisReference. 'Independent' was a one-off concept that was unique to Typescript (and isn't referenced in the spec), so I think the new name is friendlier to new readers.
2. Use `Array.every` whenever possible, since Typescript added a dependency on ES5 since the code was first written.
3. Switch to JSDoc so explanations show up nicely in editors.

Fixes #16930

### CODE REVIEW

File: tests/cases/conformance/types/thisType/thisTypeInFunctions3.ts [hunk 1]
```diff
@@ -0,0 +1,9 @@
+declare class Base {
+    check<TProp extends this>(prop: TProp): boolean;
+}
+
+class Test extends Base {
+    m() {
+        this.check(this);
```
Comment (sandersn) [inline tests/cases/conformance/types/thisType/thisTypeInFunctions3.ts]: previously the second `this` reference was not assignable to `prop: TProp`.

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -5104,10 +5104,14 @@ namespace ts {
...
-        // and if none of the base interfaces have a "this" type.
-        function isIndependentInterface(symbol: Symbol): boolean {
+        /**
+         * Returns true if the interface given by the symbol is free of "this" references.
+         *
+         * Specifically, the result is true if the interface itself contains no references
+         * to "this" in its body, if all base types are interfaces,
+         * and if none of the base interfaces have a "this" type.
+         */
+        function isInterfaceFreeOfThisReference(symbol: Symbol): boolean {
```
Comment (ahejlsberg) [inline src/compiler/checker.ts]: `isThislessInterface`?

Comment (andy-ms) [reply src/compiler/checker.ts]: Or name it `isInterfaceWithThisReference` and don't negate it

File: src/compiler/checker.ts [hunk 3]
```diff
@@ -5323,22 +5327,17 @@ namespace ts {
...
-        // A type is considered independent if it the any, string, number, boolean, symbol, or void keyword, a string
-        // literal type, an array with an element type that is considered independent, or a type reference that is
-        // considered independent.
-        function isIndependentType(node: TypeNode): boolean {
+        /**
+         * A type is free of this references if it's the any, string, number, boolean, symbol, or void keyword, a string
+         * literal type, an array with an element type that is free of this references, or a type reference that is
+         * free of this references.
+         */
+        function isTypeFreeOfThisReference(node: TypeNode): boolean {
```
Comment (andy-ms) [inline src/compiler/checker.ts]: I would also negate this.

File: src/compiler/checker.ts [hunk 4]
```diff
@@ -5353,54 +5352,58 @@ namespace ts {
                 case SyntaxKind.LiteralType:
                     return true;
                 case SyntaxKind.ArrayType:
-                    return isIndependentType((<ArrayTypeNode>node).elementType);
+                    return isTypeFreeOfThisReference((<ArrayTypeNode>node).elementType);
                 case SyntaxKind.TypeReference:
-                    return isIndependentTypeReference(<TypeReferenceNode>node);
+                    return isTypeReferenceFreeOfThisReference(<TypeReferenceNode>node);
```
Comment (andy-ms) [inline src/compiler/checker.ts]: `return some((node as TypeReferenceNode).typeArguments, typeHasThisReference);`, doesn't really deserve its own function.

File: src/compiler/checker.ts [hunk 5]
```diff
@@ -5353,54 +5352,58 @@ namespace ts {
                 case SyntaxKind.LiteralType:
                     return true;
                 case SyntaxKind.ArrayType:
-                    return isIndependentType((<ArrayTypeNode>node).elementType);
+                    return isTypeFreeOfThisReference((<ArrayTypeNode>node).elementType);
                 case SyntaxKind.TypeReference:
-                    return isIndependentTypeReference(<TypeReferenceNode>node);
+                    return isTypeReferenceFreeOfThisReference(<TypeReferenceNode>node);
             }
             return false;
```
Comment (andy-ms) [inline src/compiler/checker.ts]: So all other types are presumed guilty? Could you try for a complete `switch` and `Debug.fail()` in the default case?

Comment (sandersn) [reply src/compiler/checker.ts]: In a word, yes. I'm doing Definitely Typed duty this week, so I ran tests on the assert until they all passed. I had to add all the type kinds except for Mapped Types.

There may be some efficiency we could pick up here by inspecting more types &mdash; union and intersection especially &mdash; but I think it's fine to have the default case fall through to a `return false`.

File: src/compiler/checker.ts (hunk 3 above)
Comment (sandersn) [reply src/compiler/checker.ts]: I'll try that in a separate PR. I'm afraid that if I introduce any bugs with further refactoring they'll only show up as performance regressions, and I want to isolate that from any required performance regressions that result from the additional instantiations here.

**Pull request merged**